src/Databases/DatabaseReplicatedWorker.cpp

```diff
@@ -308,4 +313,18 @@ bool DatabaseReplicatedDDLWorker::canRemoveQueueEntry(const String & entry_name,
     return entry_number + logs_to_keep < max_log_ptr;
 }
 
+void DatabaseReplicatedDDLWorker::updateLogPointer(const String & processed_entry_name)
+{
+    updateMaxDDLEntryID(processed_entry_name);
+    assert(max_id.load() == parse<UInt32>(getAndSetZooKeeper()->get(fs::path(database->replica_path) / "log_ptr")));
+}
+
+UInt32 DatabaseReplicatedDDLWorker::getLogPointer() const
+{
+    /// NOTE it main not be equal to the log_ptr in zk:
+    ///  - max_id can be equal to log_ptr - 1 due to race condition (when it's updated in zk, but not updated in memory yet)
+    ///  - max_id can be greater than log_ptr, because log_ptr is not updated for failed and dummy entries
+    return max_id.load();
```

Comment:
We are okay with this because it's only used for checking the lag but maybe add a template boolean parameter called `safe`, `strict` or something which if set to true returns `log_ptr` from zookeeper, and if it's false returns from `max_id`.
That way I think it will be much clearer in the future that for some cases loading from `max_id` isn't fine.

Reply:
Yes, it's totally ok for calculating replication lag. However, we also use this method to check if some task is already processed or not
https://github.com/ClickHouse/ClickHouse/blob/4e9ec5dc2f7c26ee0a22655acfb195e94caa3c33/src/Databases/DatabaseReplicatedWorker.cpp#L211-L217
but it's ok too, because `max_id` is updated when task is completely processed, so we will not accidentally skip a task.

I don't think that we should add `safe`/`strict` flag now, because currently it's not needed and it will be a kind of dead code.